### PR TITLE
Simplify the way EthereumHost creates BlockChainSync

### LINKED ISSUE #3371: Data race in EthereumHost::sync()
```
==================
WARNING: ThreadSanitizer: data race (pid=7987)
  Write of size 8 at 0x7d500000efc8 by thread T14 (mutexes: write M1907):
    #0 void std::swap<dev::eth::BlockChainSync*>(dev::eth::BlockChainSync*&, dev::eth::BlockChainSync*&) /usr/include/c++/5/bits/move.h:186 (libethereum.so+0x000000239092)
    #1 std::unique_ptr<dev::eth::BlockChainSync, std::default_delete<dev::eth::BlockChainSync> >::reset(dev::eth::BlockChainSync*) /usr/include/c++/5/bits/unique_ptr.h:342 (libethereum.so+0x000000239092)
    #2 dev::eth::EthereumHost::sync() /home/chfast/Projects/ethereum/cpp-ethereum/libethereum/EthereumHost.cpp:298 (libethereum.so+0x000000239092)
    #3 dev::eth::EthereumHost::onPeerStatus(std::shared_ptr<dev::eth::EthereumPeer>) /home/chfast/Projects/ethereum/cpp-ethereum/libethereum/EthereumHost.cpp:309 (libethereum.so+0x0000002391dd)
    #4 dev::eth::EthereumPeer::interpret(unsigned int, dev::RLP const&) /home/chfast/Projects/ethereum/cpp-ethereum/libethereum/EthereumPeer.cpp:244 (libethereum.so+0x0000001471e8)
    #5 dev::p2p::Session::readPacket(unsigned short, dev::p2p::PacketType, dev::RLP const&) /home/chfast/Projects/ethereum/cpp-ethereum/libp2p/Session.cpp:179 (libp2p.so+0x0000000a71f9)
    #6 dev::p2p::Session::doRead()::{lambda(boost::system::error_code, unsigned long)#1}::operator()(boost::system::error_code, unsigned long) const::{lambda(boost::system::error_code, unsigned long)#1}::operator()(boost::system::error_code, unsigned long) const <null> (libp2p.so+0x0000000a7bfc)
    #7 operator() /usr/include/boost/asio/impl/read.hpp:282 (libp2p.so+0x0000000a8430)
    #8 operator() /usr/include/boost/asio/detail/bind_handler.hpp:127 (libp2p.so+0x0000000a9ec8)
    #9 asio_handler_invoke<boost::asio::detail::binder2<boost::asio::detail::read_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp>, boost::asio::mutable_buffers_1, boost::asio::detail::transfer_all_t, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> >, boost::system::error_code, long unsigned int> > /usr/include/boost/asio/handler_invoke_hook.hpp:69 (libp2p.so+0x0000000a9ec8)
    #10 invoke<boost::asio::detail::binder2<boost::asio::detail::read_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp>, boost::asio::mutable_buffers_1, boost::asio::detail::transfer_all_t, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> >, boost::system::error_code, long unsigned int>, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> > /usr/include/boost/asio/detail/handler_invoke_helpers.hpp:37 (libp2p.so+0x0000000a9ec8)
    #11 asio_handler_invoke<boost::asio::detail::binder2<boost::asio::detail::read_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp>, boost::asio::mutable_buffers_1, boost::asio::detail::transfer_all_t, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> >, boost::system::error_code, long unsigned int>, boost::asio::basic_stream_socket<boost::asio::ip::tcp>, boost::asio::mutable_buffers_1, boost::asio::detail::transfer_all_t, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> > /usr/include/boost/asio/impl/read.hpp:502 (libp2p.so+0x0000000a9ec8)
    #12 invoke<boost::asio::detail::binder2<boost::asio::detail::read_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp>, boost::asio::mutable_buffers_1, boost::asio::detail::transfer_all_t, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> >, boost::system::error_code, long unsigned int>, boost::asio::detail::read_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp>, boost::asio::mutable_buffers_1, boost::asio::detail::transfer_all_t, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> > > /usr/include/boost/asio/detail/handler_invoke_helpers.hpp:37 (libp2p.so+0x0000000a9ec8)
    #13 do_complete /usr/include/boost/asio/detail/reactive_socket_recv_op.hpp:110 (libp2p.so+0x0000000a9ec8)
    #14 boost::asio::detail::task_io_service_operation::complete(boost::asio::detail::task_io_service&, boost::system::error_code const&, unsigned long) /usr/include/boost/asio/detail/task_io_service_operation.hpp:38 (libp2p.so+0x0000000bc336)
    #15 boost::asio::detail::task_io_service::do_run_one(boost::asio::detail::scoped_lock<boost::asio::detail::posix_mutex>&, boost::asio::detail::task_io_service_thread_info&, boost::system::error_code const&) /usr/include/boost/asio/detail/impl/task_io_service.ipp:372 (libp2p.so+0x0000000bc336)
    #16 boost::asio::detail::task_io_service::run(boost::system::error_code&) /usr/include/boost/asio/detail/impl/task_io_service.ipp:149 (libp2p.so+0x0000000bc336)
    #17 boost::asio::io_service::run() /usr/include/boost/asio/impl/io_service.ipp:59 (libp2p.so+0x00000010e35c)
    #18 dev::p2p::Host::doWork() /home/chfast/Projects/ethereum/cpp-ethereum/libp2p/Host.cpp:771 (libp2p.so+0x00000010e35c)
    #19 dev::Worker::workLoop() /home/chfast/Projects/ethereum/cpp-ethereum/libdevcore/Worker.cpp:120 (libdevcore.so+0x000000069240)
    #20 dev::Worker::startWorking()::{lambda()#1}::operator()() const [clone .constprop.230] <null> (libdevcore.so+0x00000006a541)
    #21 _M_invoke<> /usr/include/c++/5/functional:1531 (libdevcore.so+0x00000006c84d)
    #22 operator() /usr/include/c++/5/functional:1520 (libdevcore.so+0x00000006c84d)
    #23 _M_run /usr/include/c++/5/thread:115 (libdevcore.so+0x00000006c84d)
    #24 <null> <null> (libstdc++.so.6+0x0000000b8c7f)

  Previous read of size 8 at 0x7d500000efc8 by thread T15:
    #0 std::unique_ptr<dev::eth::BlockChainSync, std::default_delete<dev::eth::BlockChainSync> >::get() const /usr/include/c++/5/bits/unique_ptr.h:305 (libethereum.so+0x000000233960)
    #1 std::unique_ptr<dev::eth::BlockChainSync, std::default_delete<dev::eth::BlockChainSync> >::operator bool() const /usr/include/c++/5/bits/unique_ptr.h:319 (libethereum.so+0x000000233960)
    #2 dev::eth::EthereumHost::isSyncing() const /home/chfast/Projects/ethereum/cpp-ethereum/libethereum/EthereumHost.cpp:419 (libethereum.so+0x000000233960)
    #3 dev::eth::EthereumHost::doWork() /home/chfast/Projects/ethereum/cpp-ethereum/libethereum/EthereumHost.cpp:102 (libethereum.so+0x00000023d97a)
    #4 dev::Worker::workLoop() /home/chfast/Projects/ethereum/cpp-ethereum/libdevcore/Worker.cpp:120 (libdevcore.so+0x000000069240)
    #5 dev::Worker::startWorking()::{lambda()#1}::operator()() const [clone .constprop.230] <null> (libdevcore.so+0x00000006a541)
    #6 _M_invoke<> /usr/include/c++/5/functional:1531 (libdevcore.so+0x00000006c84d)
    #7 operator() /usr/include/c++/5/functional:1520 (libdevcore.so+0x00000006c84d)
    #8 _M_run /usr/include/c++/5/thread:115 (libdevcore.so+0x00000006c84d)
    #9 <null> <null> (libstdc++.so.6+0x0000000b8c7f)

  Location is heap block of size 480 at 0x7d500000ee00 allocated by main thread:
    #0 operator new(unsigned long) <null> (libtsan.so.0+0x000000025a33)
    #1 __gnu_cxx::new_allocator<std::_Sp_counted_ptr_inplace<dev::eth::EthereumHost, std::allocator<dev::eth::EthereumHost>, (__gnu_cxx::_Lock_policy)2> >::allocate(unsigned long, void const*) /usr/include/c++/5/ext/new_allocator.h:104 (libethereum.so+0x0000001815ef)
    #2 std::allocator_traits<std::allocator<std::_Sp_counted_ptr_inplace<dev::eth::EthereumHost, std::allocator<dev::eth::EthereumHost>, (__gnu_cxx::_Lock_policy)2> > >::allocate(std::allocator<std::_Sp_counted_ptr_inplace<dev::eth::EthereumHost, std::allocator<dev::eth::EthereumHost>, (__gnu_cxx::_Lock_policy)2> >&, unsigned long) /usr/include/c++/5/bits/alloc_traits.h:491 (libethereum.so+0x0000001815ef)
    #3 std::__allocated_ptr<std::allocator<std::_Sp_counted_ptr_inplace<dev::eth::EthereumHost, std::allocator<dev::eth::EthereumHost>, (__gnu_cxx::_Lock_policy)2> > > std::__allocate_guarded<std::allocator<std::_Sp_counted_ptr_inplace<dev::eth::EthereumHost, std::allocator<dev::eth::EthereumHost>, (__gnu_cxx::_Lock_policy)2> > >(std::allocator<std::_Sp_counted_ptr_inplace<dev::eth::EthereumHost, std::allocator<dev::eth::EthereumHost>, (__gnu_cxx::_Lock_policy)2> >&) /usr/include/c++/5/bits/allocated_ptr.h:102 (libethereum.so+0x0000001815ef)
    #4 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<dev::eth::EthereumHost, std::allocator<dev::eth::EthereumHost>, dev::eth::BlockChain&, dev::OverlayDB&, dev::eth::TransactionQueue&, dev::eth::BlockQueue&, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>&>(std::_Sp_make_shared_tag, dev::eth::EthereumHost*, std::allocator<dev::eth::EthereumHost> const&, dev::eth::BlockChain&, dev::OverlayDB&, dev::eth::TransactionQueue&, dev::eth::BlockQueue&, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>&) /usr/include/c++/5/bits/shared_ptr_base.h:615 (libethereum.so+0x0000001815ef)
    #5 std::__shared_ptr<dev::eth::EthereumHost, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<dev::eth::EthereumHost>, dev::eth::BlockChain&, dev::OverlayDB&, dev::eth::TransactionQueue&, dev::eth::BlockQueue&, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>&>(std::_Sp_make_shared_tag, std::allocator<dev::eth::EthereumHost> const&, dev::eth::BlockChain&, dev::OverlayDB&, dev::eth::TransactionQueue&, dev::eth::BlockQueue&, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>&) /usr/include/c++/5/bits/shared_ptr_base.h:1097 (libethereum.so+0x0000001815ef)
    #6 std::shared_ptr<dev::eth::EthereumHost>::shared_ptr<std::allocator<dev::eth::EthereumHost>, dev::eth::BlockChain&, dev::OverlayDB&, dev::eth::TransactionQueue&, dev::eth::BlockQueue&, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>&>(std::_Sp_make_shared_tag, std::allocator<dev::eth::EthereumHost> const&, dev::eth::BlockChain&, dev::OverlayDB&, dev::eth::TransactionQueue&, dev::eth::BlockQueue&, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>&) /usr/include/c++/5/bits/shared_ptr.h:319 (libethereum.so+0x0000001815ef)
    #7 std::shared_ptr<dev::eth::EthereumHost> std::allocate_shared<dev::eth::EthereumHost, std::allocator<dev::eth::EthereumHost>, dev::eth::BlockChain&, dev::OverlayDB&, dev::eth::TransactionQueue&, dev::eth::BlockQueue&, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>&>(std::allocator<dev::eth::EthereumHost> const&, dev::eth::BlockChain&, dev::OverlayDB&, dev::eth::TransactionQueue&, dev::eth::BlockQueue&, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>&) /usr/include/c++/5/bits/shared_ptr.h:620 (libethereum.so+0x0000001815ef)
    #8 std::shared_ptr<dev::eth::EthereumHost> std::make_shared<dev::eth::EthereumHost, dev::eth::BlockChain&, dev::OverlayDB&, dev::eth::TransactionQueue&, dev::eth::BlockQueue&, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>&>(dev::eth::BlockChain&, dev::OverlayDB&, dev::eth::TransactionQueue&, dev::eth::BlockQueue&, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>&) /usr/include/c++/5/bits/shared_ptr.h:636 (libethereum.so+0x0000001815ef)
    #9 dev::eth::Client::init(dev::p2p::Host*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, dev::WithExisting, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>) /home/chfast/Projects/ethereum/cpp-ethereum/libethereum/Client.cpp:110 (libethereum.so+0x0000001815ef)
    #10 dev::eth::Client::Client(dev::eth::ChainParams const&, int, dev::p2p::Host*, std::shared_ptr<dev::eth::GasPricer>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, dev::WithExisting, dev::eth::TransactionQueue::Limits const&) /home/chfast/Projects/ethereum/cpp-ethereum/libethereum/Client.cpp:76 (libethereum.so+0x000000183596)
    #11 dev::eth::EthashClient::EthashClient(dev::eth::ChainParams const&, int, dev::p2p::Host*, std::shared_ptr<dev::eth::GasPricer>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, dev::WithExisting, dev::eth::TransactionQueue::Limits const&) /home/chfast/Projects/ethereum/cpp-ethereum/libethashseal/EthashClient.cpp:54 (libethashseal.so+0x0000000a5302)
    #12 dev::WebThreeDirect::WebThreeDirect(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, dev::eth::ChainParams const&, dev::WithExisting, std::set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, dev::p2p::NetworkPreferences const&, dev::vector_ref<unsigned char const>, bool) /home/chfast/Projects/ethereum/cpp-ethereum/libwebthree/WebThree.cpp:63 (libwebthree.so+0x0000000510ad)
    #13 main /home/chfast/Projects/ethereum/cpp-ethereum/eth/main.cpp:1047 (eth+0x000000443e2c)

  Mutex M1907 (0x7d500000ef78) created at:
    #0 pthread_mutex_lock <null> (libtsan.so.0+0x000000037286)
    #1 __gthread_mutex_lock /usr/include/x86_64-linux-gnu/c++/5/bits/gthr-default.h:748 (libethereum.so+0x0000002391cd)
    #2 __gthread_recursive_mutex_lock /usr/include/x86_64-linux-gnu/c++/5/bits/gthr-default.h:810 (libethereum.so+0x0000002391cd)
    #3 std::recursive_mutex::lock() /usr/include/c++/5/mutex:176 (libethereum.so+0x0000002391cd)
    #4 std::lock_guard<std::recursive_mutex>::lock_guard(std::recursive_mutex&) /usr/include/c++/5/mutex:386 (libethereum.so+0x0000002391cd)
    #5 dev::eth::EthereumHost::onPeerStatus(std::shared_ptr<dev::eth::EthereumPeer>) /home/chfast/Projects/ethereum/cpp-ethereum/libethereum/EthereumHost.cpp:308 (libethereum.so+0x0000002391cd)
    #6 dev::eth::EthereumPeer::interpret(unsigned int, dev::RLP const&) /home/chfast/Projects/ethereum/cpp-ethereum/libethereum/EthereumPeer.cpp:244 (libethereum.so+0x0000001471e8)
    #7 dev::p2p::Session::readPacket(unsigned short, dev::p2p::PacketType, dev::RLP const&) /home/chfast/Projects/ethereum/cpp-ethereum/libp2p/Session.cpp:179 (libp2p.so+0x0000000a71f9)
    #8 dev::p2p::Session::doRead()::{lambda(boost::system::error_code, unsigned long)#1}::operator()(boost::system::error_code, unsigned long) const::{lambda(boost::system::error_code, unsigned long)#1}::operator()(boost::system::error_code, unsigned long) const <null> (libp2p.so+0x0000000a7bfc)
    #9 operator() /usr/include/boost/asio/impl/read.hpp:282 (libp2p.so+0x0000000a8430)
    #10 operator() /usr/include/boost/asio/detail/bind_handler.hpp:127 (libp2p.so+0x0000000a9ec8)
    #11 asio_handler_invoke<boost::asio::detail::binder2<boost::asio::detail::read_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp>, boost::asio::mutable_buffers_1, boost::asio::detail::transfer_all_t, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> >, boost::system::error_code, long unsigned int> > /usr/include/boost/asio/handler_invoke_hook.hpp:69 (libp2p.so+0x0000000a9ec8)
    #12 invoke<boost::asio::detail::binder2<boost::asio::detail::read_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp>, boost::asio::mutable_buffers_1, boost::asio::detail::transfer_all_t, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> >, boost::system::error_code, long unsigned int>, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> > /usr/include/boost/asio/detail/handler_invoke_helpers.hpp:37 (libp2p.so+0x0000000a9ec8)
    #13 asio_handler_invoke<boost::asio::detail::binder2<boost::asio::detail::read_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp>, boost::asio::mutable_buffers_1, boost::asio::detail::transfer_all_t, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> >, boost::system::error_code, long unsigned int>, boost::asio::basic_stream_socket<boost::asio::ip::tcp>, boost::asio::mutable_buffers_1, boost::asio::detail::transfer_all_t, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> > /usr/include/boost/asio/impl/read.hpp:502 (libp2p.so+0x0000000a9ec8)
    #14 invoke<boost::asio::detail::binder2<boost::asio::detail::read_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp>, boost::asio::mutable_buffers_1, boost::asio::detail::transfer_all_t, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> >, boost::system::error_code, long unsigned int>, boost::asio::detail::read_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp>, boost::asio::mutable_buffers_1, boost::asio::detail::transfer_all_t, dev::p2p::Session::doRead()::<lambda(boost::system::error_code, std::size_t)>::<lambda(boost::system::error_code, std::size_t)> > > /usr/include/boost/asio/detail/handler_invoke_helpers.hpp:37 (libp2p.so+0x0000000a9ec8)
    #15 do_complete /usr/include/boost/asio/detail/reactive_socket_recv_op.hpp:110 (libp2p.so+0x0000000a9ec8)
    #16 boost::asio::detail::task_io_service_operation::complete(boost::asio::detail::task_io_service&, boost::system::error_code const&, unsigned long) /usr/include/boost/asio/detail/task_io_service_operation.hpp:38 (libp2p.so+0x0000000bc336)
    #17 boost::asio::detail::task_io_service::do_run_one(boost::asio::detail::scoped_lock<boost::asio::detail::posix_mutex>&, boost::asio::detail::task_io_service_thread_info&, boost::system::error_code const&) /usr/include/boost/asio/detail/impl/task_io_service.ipp:372 (libp2p.so+0x0000000bc336)
    #18 boost::asio::detail::task_io_service::run(boost::system::error_code&) /usr/include/boost/asio/detail/impl/task_io_service.ipp:149 (libp2p.so+0x0000000bc336)
    #19 boost::asio::io_service::run() /usr/include/boost/asio/impl/io_service.ipp:59 (libp2p.so+0x00000010e35c)
    #20 dev::p2p::Host::doWork() /home/chfast/Projects/ethereum/cpp-ethereum/libp2p/Host.cpp:771 (libp2p.so+0x00000010e35c)
    #21 dev::Worker::workLoop() /home/chfast/Projects/ethereum/cpp-ethereum/libdevcore/Worker.cpp:120 (libdevcore.so+0x000000069240)
    #22 dev::Worker::startWorking()::{lambda()#1}::operator()() const [clone .constprop.230] <null> (libdevcore.so+0x00000006a541)
    #23 _M_invoke<> /usr/include/c++/5/functional:1531 (libdevcore.so+0x00000006c84d)
    #24 operator() /usr/include/c++/5/functional:1520 (libdevcore.so+0x00000006c84d)
    #25 _M_run /usr/include/c++/5/thread:115 (libdevcore.so+0x00000006c84d)
    #26 <null> <null> (libstdc++.so.6+0x0000000b8c7f)

  Thread T14 'p2p' (tid=8006, running) created by main thread at:
    #0 pthread_create <null> (libtsan.so.0+0x000000027577)
    #1 std::thread::_M_start_thread(std::shared_ptr<std::thread::_Impl_base>, void (*)()) <null> (libstdc++.so.6+0x0000000b8dc2)
    #2 dev::p2p::Host::start() /home/chfast/Projects/ethereum/cpp-ethereum/libp2p/Host.cpp:127 (libp2p.so+0x0000001126c8)
    #3 dev::WebThreeDirect::startNetwork() /home/chfast/Projects/ethereum/cpp-ethereum/eth/../libwebthree/WebThree.h:208 (eth+0x0000004451c1)
    #4 main /home/chfast/Projects/ethereum/cpp-ethereum/eth/main.cpp:1219 (eth+0x0000004451c1)

  Thread T15 'ethsync' (tid=8007, running) created by thread T14 at:
    #0 pthread_create <null> (libtsan.so.0+0x000000027577)
    #1 std::thread::_M_start_thread(std::shared_ptr<std::thread::_Impl_base>, void (*)()) <null> (libstdc++.so.6+0x0000000b8dc2)
    #2 dev::eth::EthereumHost::onStarting() /home/chfast/Projects/ethereum/cpp-ethereum/libethereum/EthereumHost.h:117 (libethereum.so+0x00000023ec6d)
    #3 dev::p2p::Host::startedWorking() /home/chfast/Projects/ethereum/cpp-ethereum/libp2p/Host.cpp:738 (libp2p.so+0x00000012564a)
    #4 dev::Worker::startWorking()::{lambda()#1}::operator()() const [clone .constprop.230] <null> (libdevcore.so+0x00000006a7f5)
    #5 _M_invoke<> /usr/include/c++/5/functional:1531 (libdevcore.so+0x00000006c84d)
    #6 operator() /usr/include/c++/5/functional:1520 (libdevcore.so+0x00000006c84d)
    #7 _M_run /usr/include/c++/5/thread:115 (libdevcore.so+0x00000006c84d)
    #8 <null> <null> (libstdc++.so.6+0x0000000b8c7f)

SUMMARY: ThreadSanitizer: data race /usr/include/c++/5/bits/move.h:186 void std::swap<dev::eth::BlockChainSync*>(dev::eth::BlockChainSync*&, dev::eth::BlockChainSync*&)
==================
```

### PR DESCRIPTION
**Pull request by chfast**
Fixes #3371.

### CODE REVIEW

File: libethereum/EthereumHost.cpp [hunk 1]
```diff
@@ -58,6 +58,9 @@ EthereumHost::EthereumHost(BlockChain const& _ch, OverlayDB const& _db, Transact
 	m_bq		(_bq),
 	m_networkId	(_networkId)
 {
+	// TODO: Composition would be better. Left like that to avoid initialization
+	//       issues as BlockChainSync accesses other EthereumHost members.
+	m_sync.reset(new BlockChainSync(*this));
```
Comment (chriseth) [inline libethereum/EthereumHost.cpp]: `m_sync = make_shared<BlockChainSync>(*this);`?

**Comment by chfast** [comment]
I pushed 2 small fixes here. Reviewing by commits might be easier.

File: libethereum/EthereumHost.cpp [hunk 2]
```diff
@@ -120,21 +120,6 @@ void EthereumHost::doWork()
 		foreachPeer([](std::shared_ptr<EthereumPeer> _p) { _p->tick(); return true; });
 	}
 
-	if (m_syncStart)
```
Comment (chriseth) [inline libethereum/EthereumHost.cpp]: Can you explain this change and the meaning of `m_syncStart` in general?

File: libethereum/EthereumHost.cpp (hunk 1 above)
Comment (chfast) [inline libethereum/EthereumHost.cpp]: This is unique_ptr. If you want to replace missing make_unique I think this is bad idea. make_shared allocates more memory than needed.

File: libethereum/EthereumHost.cpp (hunk 2 above)
Comment (chfast) [inline libethereum/EthereumHost.cpp]: It was like that:
1. sync() tries to find PV63 peer.
2. If failed, it sets the m_syncStart and does not create BlockChainSync.
3. In the reset function here, if 10s after m_syncStart the BlockChainSync is created.

I just tried to create the BlockChainSync straight away it I can't see any difference. Maybe it had more sense when there was PV61/62 support.

Comment (winsvega) [inline libethereum/EthereumHost.cpp]: whats with windows appveyour?

**Comment by winsvega** [comment]
whats with windows appveyour?

**Comment by chfast** [comment]
It run out of time.

File: libethereum/EthereumHost.cpp (hunk 1 above)
Comment (chriseth) [inline libethereum/EthereumHost.cpp]: Oh I'm sorry, yes, it's fine.

**Pull request merged**